test/server/http_time_tracking_filter_integration_test.cc

```diff
@@ -0,0 +1,147 @@
+const std::string kDefaultProtoFragment = fmt::format(
+    "emit_previous_request_delta_in_response_header: \"{}\"", kLatencyResponseHeaderName);
+const std::string kProtoConfigTemplate = R"EOF(
+name: time-tracking
+typed_config:
+  "@type": type.googleapis.com/nighthawk.server.ResponseOptions
+  {}
+)EOF";
+
+class HttpTimeTrackingIntegrationTest
+    : public Envoy::HttpIntegrationTest,
+      public testing::TestWithParam<Envoy::Network::Address::IpVersion> {
+protected:
+  HttpTimeTrackingIntegrationTest()
+      : HttpIntegrationTest(Envoy::Http::CodecClient::Type::HTTP1, GetParam()),
+        request_headers_({{":method", "GET"}, {":path", "/"}, {":authority", "host"}}) {}
+
+  // We don't override SetUp(): tests in this file will call setup() instead to avoid having to
+  // create a fixture per filter configuration.
+  void setup(const std::string& config) {
+    config_helper_.addFilter(config);
+    HttpIntegrationTest::initialize();
+  }
+
+  // Fetches a response with request-level configuration set in the request header.
+  Envoy::IntegrationStreamDecoderPtr getResponse(absl::string_view request_level_config,
+                                                 bool setup_for_upstream_request = true) {
+    Envoy::Http::TestRequestHeaderMapImpl request_headers = request_headers_;
+    request_headers.setCopy(Nighthawk::Server::TestServer::HeaderNames::get().TestServerConfig,
+                            request_level_config);
+    return getResponse(request_headers, setup_for_upstream_request);
+  }
+
+  // Fetches a response with the default request headers, expecting the fake upstream to supply
+  // the response.
+  Envoy::IntegrationStreamDecoderPtr getResponse() { return getResponse(request_headers_); }
+
+  // Fetches a response using the provided request headers. When setup_for_upstream_request
+  // is true, the expectation will be that an upstream request will be needed to provide a
+  // response. If it is set to false, the extension is expected to supply the response, and
+  // no upstream request ought to occur.
+  Envoy::IntegrationStreamDecoderPtr
+  getResponse(const Envoy::Http::TestRequestHeaderMapImpl& request_headers,
+              bool setup_for_upstream_request = true) {
+    cleanupUpstreamAndDownstream();
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    Envoy::IntegrationStreamDecoderPtr response;
+    if (setup_for_upstream_request) {
+      response = sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+    } else {
+      response = codec_client_->makeHeaderOnlyRequest(request_headers);
+      response->waitForEndStream();
+    }
+    return response;
+  }
+
+  const Envoy::Http::TestRequestHeaderMapImpl request_headers_;
+};
+
+class HttpTimeTrackingFilterConfigTest : public testing::Test,
+                                         public Envoy::Event::TestUsingSimulatedTime {};
+
+// WARNING: Don't move this test. Keep it on top.
```

Comment:
This comment indicates a design problem to me.  Could we instead abstract this to a function that is called before every test?  And then, if it's necessary for it not be called multiple times, verify that there instead?

Reply:
The problem is that the static persists across tests. So only the first call from this compilation unit will truly excite the behaviour we are after consolidating in the test here, regardless of which test triggers it. I fully agree it's a code smell, but resolving it either requires  introducing test-specific code to reset the static to its initial value. I'll be in touch on Slack to discuss this some more -- I'm not sure I grasp what you propose here so I think I need some more context -- I'm surely open to ideas to improve upon this.

Reply:
I think I resolved this code smell in 7774e2d; indeed design was flawed. With those changes, tests no longer need to be order-dependent, as we can control instantiation, the static/singleton wasn't needed. Verified that with gtest's `--gtest_shuffle` flag .